synfig-studio/src/gui/dialogs/dialog_pluginmanager.cpp

```diff
@@ -0,0 +1,298 @@
+            std::string output_path ;
+            std::string plugin_metadata_file;
+            std::string plugin_name ;
+
+            if (!zip_fs->open(plugin_file_dialog.get_filename())) {
+                synfig::error("Failed to open zip file: " + zip_filename);
+                return;
+            }
+            zip_fs->directory_scan("", files);
+            for (const auto& file : files) {
+                if (file == "plugin.xml") {
+                    plugin_metadata_file = file;
+                    break;
+                } else if(zip_fs->is_directory(file)) {
+                    std::vector<std::string> child_files;
+                    if (zip_fs->directory_scan(file, child_files)) {
+                    auto it = std::find(child_files.begin(), child_files.end(), "plugin.xml");
+                    if (it != child_files.end()) {
+                        plugin_metadata_file = file + "/" + *it;
+                    }
+                    if (!plugin_metadata_file.empty())
+                        break;
+                    }
+                }
+            }
+            if (plugin_metadata_file.empty() || !zip_fs->is_file(plugin_metadata_file)) {
+                synfig::error("Failed to find plugin.xml in zip file: " + zip_filename);
+                return;
+            }
+            plugin_name = extract_plugin_name(*zip_fs->get_read_stream(plugin_metadata_file));
+            output_path = ((path_to_user_plugins / plugin_name).add_suffix("/")).u8string();
+            if (native_fs->is_exists(output_path) && native_fs->is_directory(output_path)) {
+                message_dialog.set_message(_("Plugin already exists. Do you want to overwrite it?"));
+                int response = message_dialog.run();
+                if (response != Gtk::RESPONSE_OK) {
+                    message_dialog.close();
+                    return;
+                }
+                auto it = std::find_if(App::plugin_manager.plugins().begin(), App::plugin_manager.plugins().end(), [plugin_name](const Plugin& plugin) { return plugin.name.fallback() == plugin_name; });
+                if (it != App::plugin_manager.plugins().end()) {
+                    App::plugin_manager.remove_plugin(it->id);
+                }
+                native_fs->remove_recursive(output_path);
+                message_dialog.close();
+            }
+            if (native_fs->is_file(output_path)) {
+                if (!native_fs->file_remove(output_path)) {
+                    synfig::error("Failed to remove file: " + output_path);
+                    return;
+                }
+            }
+            if (native_fs->directory_create(output_path)) {
+                if (plugin_metadata_file.find("/") == std::string::npos) {
+                    for(const auto& file : files) {
+                        zip_fs->copy_recursive(zip_fs,  file, native_fs, output_path + file);
+                    }
```

Comment:
Why do you  want to extract all files from zip here?

Reply:
The zip contains the files for plugin, and when user installs it, those files will be installed in the plugins directory for the user.

(This should be highly obvious, am I missing something from the question?)

Reply:
You plan to support two plugin zip file layouts:
1. A 'plain' zip file, where 
2. A zip file that contains a folder that contains the metadata plugin.xml and all its other files

The 'its other files' may include folders too.

Suppose a badly structured zip file with this layout:
- a folder called `skel` (in the zip root), that contains:
  - plugin.xml
  - skel.py
- another folder (in the zip root) named `trash`
  - trash.txt
  - whyonhell.txt

If I read the code correctly, your code will extract both folders in the plugin folder you created. You should extract the `skel` folder contents only. Or am I wrong?

Reply:
There are some corrections in your understanding. 
The above pointed code is only executed when if the "plugin.xml" file is at root level. If the "plugin.xml" is at sub directory then the below code is executed:
https://github.com/synfig/synfig/blob/e985c198cc1d53814c62e55e43fe8ef738367a4b/synfig-studio/src/gui/dialogs/dialog_pluginmanager.cpp#L283-L285

(Which requires some correction, because it assumes that the zip file contains only one directory)